File: mcp-server/internal/domains/resourcehandler.go
package domains

import (
	"context"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// VersionedFetchFunc is the signature for domain Fetch functions that accept
// an additional version parameter (e.g. skills).
type VersionedFetchFunc func(ctx context.Context, serverAddr, org, slug, version string) (string, error)

// NewResourceHandler creates a standard MCP resource handler that parses
// org and slug from the request URI and delegates to fetchFn.
func NewResourceHandler(fetchFn FetchFunc, serverAddr, domainName string) mcp.ResourceHandler {
	return func(ctx context.Context, req *mcp.ReadResourceRequest) (*mcp.ReadResourceResult, error) {
		org, slug, err := ParseResourceURI(req.Params.URI)
		if err != nil {
			return nil, fmt.Errorf("%s resource: %w", domainName, err)
		}
		text, err := fetchFn(ctx, serverAddr, org, slug)
		if err != nil {
			return nil, err
		}
		return ResourceResult(req.Params.URI, text), nil
	}
}

// NewVersionedResourceHandler creates an MCP resource handler that parses
// org, slug, and version from the request URI and delegates to fetchFn.
func NewVersionedResourceHandler(fetchFn VersionedFetchFunc, serverAddr, domainName string) mcp.ResourceHandler {
	return func(ctx context.Context, req *mcp.ReadResourceRequest) (*mcp.ReadResourceResult, error) {
		org, slug, version, err := ParseVersionedResourceURI(req.Params.URI)
		if err != nil {
			return nil, fmt.Errorf("%s versioned resource: %w", domainName, err)
		}
		text, err := fetchFn(ctx, serverAddr, org, slug, version)
		if err != nil {
			return nil, err
		}
		return ResourceResult(req.Params.URI, text), nil
	}
}

// ResourceResult constructs a ReadResourceResult with a single JSON text
// content entry.
func ResourceResult(uri, text string) *mcp.ReadResourceResult {
	return &mcp.ReadResourceResult{
		Contents: []*mcp.ResourceContents{{
			URI:      uri,
			MIMEType: "application/json",
			Text:     text,
		}},
	}
}

File: mcp-server-ts/src/domains/resourcehandler.ts
// Registration helpers for MCP resource templates.
// Go parity: mcp-server/internal/domains/resourcehandler.go.
//
// Every Stigmer resource template resolves the per-request credential the same
// way tools do (resolveToken over extra.authInfo), parses org/slug[/version]
// from the request URI, delegates to the domain fetch, and returns a single
// application/json content entry. These helpers keep each domain's resources.ts
// a one-liner and ensure the auth + result shape stay identical across domains.

import { McpServer, ResourceTemplate } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { ReadResourceResult } from "@modelcontextprotocol/sdk/types.js";

import { resolveToken, type BackendTarget } from "./client";
import { parseResourceURI, parseVersionedResourceURI } from "./resourceuri";

/** Domain read used by a non-versioned resource template. */
export type ResourceFetch = (
  serverAddress: string,
  token: string,
  org: string,
  slug: string,
) => Promise<string>;

/** Domain read used by a versioned resource template ("" version means latest). */
export type VersionedResourceFetch = (
  serverAddress: string,
  token: string,
  org: string,
  slug: string,
  version: string,
) => Promise<string>;

/** Options shared by both registration helpers. */
interface ResourceOptions {
  /** Stable resource name surfaced in resources/templates/list (e.g. stigmer_agent). */
  readonly name: string;
  /** Human title surfaced to clients (e.g. "Stigmer Agent"). */
  readonly title: string;
  /** Human description surfaced to clients. */
  readonly description: string;
  /** RFC-6570 URI template, e.g. stigmer://agents/{org}/{slug}. */
  readonly template: string;
}

/** Register a stigmer://{authority}/{org}/{slug} resource template. */
export function registerResource(
  server: McpServer,
  target: BackendTarget,
  opts: ResourceOptions & { readonly fetch: ResourceFetch },
): void {
  server.registerResource(
    opts.name,
    new ResourceTemplate(opts.template, { list: undefined }),
    { title: opts.title, description: opts.description, mimeType: "application/json" },
    async (uri, _vars, extra) => {
      const { org, slug } = parseResourceURI(uri.href);
      const text = await opts.fetch(target.serverAddress, resolveToken(extra, target.apiKey), org, slug);
      return jsonResource(uri.href, text);
    },
  );
}

/** Register a stigmer://{authority}/{org}/{slug}[/{version}] resource template. */
export function registerVersionedResource(
  server: McpServer,
  target: BackendTarget,
  opts: ResourceOptions & { readonly fetch: VersionedResourceFetch },
): void {
  server.registerResource(
    opts.name,
    new ResourceTemplate(opts.template, { list: undefined }),
    { title: opts.title, description: opts.description, mimeType: "application/json" },
    async (uri, _vars, extra) => {
      const { org, slug, version } = parseVersionedResourceURI(uri.href);
      const text = await opts.fetch(
        target.serverAddress,
        resolveToken(extra, target.apiKey),
        org,
        slug,
        version,
      );
      return jsonResource(uri.href, text);
    },
  );
}

/** Wrap fetched JSON text into the single-entry ReadResourceResult tools expect. */
function jsonResource(uri: string, text: string): ReadResourceResult {
  return { contents: [{ uri, mimeType: "application/json", text }] };
}
